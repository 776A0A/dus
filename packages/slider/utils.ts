export function getTranslateX(el: HTMLElement) {
	return (
		el.style.transform.match(/translate3d\((.+)(px|%),\s*.+,\s*.+\)/)?.[
			1
		] ?? '0'
	);
}

export function transform(el: HTMLElement, x: number, scale: number) {
	el.style.transform = `translate3d(${x}%, 0, 0) scale(${scale})`;
}

export function transition(el: HTMLElement, transition: string) {
	el.style.transition = transition;
}
