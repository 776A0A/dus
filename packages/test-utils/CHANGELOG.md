# @dus/test-utils

## 1.0.1

### Patch Changes

- 修复发布了私有包

## 1.0.0

### Major Changes

- publish first release.
