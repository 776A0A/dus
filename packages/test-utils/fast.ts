import { vi } from 'vitest';

export async function fast<F extends () => Promise<any>>(callback: F): Promise<
	Awaited<ReturnType<F>>
> {
	vi.useFakeTimers();

	const result: Awaited<ReturnType<F>> = await callback();

	vi.runAllTimers();
	vi.useRealTimers();

	return result;
}
