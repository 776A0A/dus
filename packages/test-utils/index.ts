export * from './asyncMount';
export * from './fast';
