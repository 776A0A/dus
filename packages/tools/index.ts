export * from './src/array';
export * from './src/arrayBufferToBase64';
export * from './src/base64ToArrayBuffer';
export * from './src/base64ToFile';
export * from './src/base64ToImage';
export * from './src/browsers';
export * from './src/buildFormData';
export * from './src/md5';
export * from './src/callInSafe';
export * from './src/callAsyncInSafe';
export * from './src/cleanNullable';
export * from './src/convertNames';
export * from './src/copy';
export * from './src/diff';
export * from './src/download';
export * from './src/escapeDanger';
export * from './src/genId';
export * from './src/getFilenameFromHeader';
export * from './src/getString';
export * from './src/getStringLength';
export * from './src/imageFileToBase64';
export * from './src/imageElToBase64';
export * from './src/type';
export * from './src/isHttps';
export * from './src/logError';
export * from './src/normalizeUrl';
export * from './src/padPrefix';
export * from './src/preloadImages';
export * from './src/probability';
export * from './src/promise';
export * from './src/reload';
export * from './src/reloadPage';
export * from './src/removeSequentialSpace';
export * from './src/repeatable';
export * from './src/concurrent';
export * from './src/ext';
export * from './src/tinyCrypto';
export * from './src/sleep';
export * from './src/splitScreen';
export * from './src/texts';
export * from './src/urlToFile';
export * from './src/urlToImageEl';
