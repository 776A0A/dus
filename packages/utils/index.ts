export * from './src/array';
export * from './src/arrayBufferToBase64';
export * from './src/base64ToArrayBuffer';
export * from './src/base64ToFile';
export * from './src/base64ToImage';
export * from './src/browsers';
export * from './src/buildFormData';
export * from './src/calcFileMd5';
export * from './src/callWithErrorCatch';
export * from './src/callWithErrorCatchAsync';
export * from './src/cleanObject';
export * from './src/colorUtil';
export * from './src/convertObjName';
export * from './src/copy';
export * from './src/downloadFile';
export * from './src/escapeHTML';
export * from './src/generateId';
export * from './src/getDiff';
export * from './src/getFilenameFromHeader';
export * from './src/getString';
export * from './src/getStringLength';
export * from './src/imageFileToBase64';
export * from './src/imageToBase64';
export * from './src/is';
export * from './src/isHttps';
export * from './src/logError';
export * from './src/normalizeUrl';
export * from './src/padPrefix';
export * from './src/preloadImage';
export * from './src/promise';
export * from './src/reload';
export * from './src/reloadPage';
export * from './src/removeExtraSpaces';
export * from './src/repeatAction';
export * from './src/requestParallel';
export * from './src/resolveFileExt';
export * from './src/simpleCrypto';
export * from './src/sleep';
export * from './src/splitScreen';
export * from './src/texts';
export * from './src/urlToFile';
export * from './src/urlToImage';
