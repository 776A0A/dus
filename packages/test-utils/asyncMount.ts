import { mount } from '@vue/test-utils';
import { nextTick } from 'vue';
import merge from 'lodash.merge';
import { injectBrowserEnv } from './injectBrowserEnv';

export async function asyncMount<C = Parameters<typeof mount>[0]>(
	component: Partial<C>,
	props?: Parameters<typeof mount>[1],
) {
	injectBrowserEnv();

	if (props) {
		merge(props, { attachTo: 'body' });
	}

	const wrapper = mount(component as C, props);

	await nextTick();

	return wrapper;
}
