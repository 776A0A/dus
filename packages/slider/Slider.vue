<template>
  <div
    ref="sliderContainerEl"
    class="flex relative overflow-hidden slider-container"
    @mouseover="pause"
    @mouseleave="autoplay && run()"
  >
    <div
      v-for="(item, i) in activePool"
      class="slider-item will-change-transform"
      :class="{ 'slider-item__active': i === 2 }"
    >
      <slot :data="item" />
    </div>
  </div>
</template>

<script lang="ts" setup>
import { array } from '@dz7/tools'
import { useEventListener } from '@vueuse/core'
import { nextTick, onMounted, reactive, ref } from 'vue'
import { getIndex, getTranslateX, transform, transition } from './utils'

const props = withDefaults(
  defineProps<{
    list?: string[]
    scale?: number
    duration?: number
    autoplay?: boolean
    toRight?: boolean
  }>(),
  { list: () => [], scale: 0.8, duration: 1000, autoplay: false }
)

const emit = defineEmits<{
  (e: 'pause'): void
  (e: 'activeChange', active: number): void
}>()

const DURATION = props.duration
const sliderContainerEl = ref<HTMLDivElement>()
const active = ref(1)
const sliderItemEls = reactive<HTMLDivElement[]>([])
const activePool = ref<string[]>([])
let timer: NodeJS.Timer
let sliding = false

useEventListener(document, 'visibilitychange', () => {
  if (document.visibilityState === 'hidden') cancel()
  else if (props.autoplay) run()
})

onMounted(async () => {
  if (!props.list.length) return

  await init()

  if (props.autoplay) run()
})

defineExpose({ slideLeft, slideRight, active, pause })

async function init() {
  fillActivePool()

  await nextTick()

  getSliderItemEls()

  reset()
}

function pause() {
  cancel()

  emit('pause')
}

function fillActivePool() {
  const images = props.list,
    length = images.length

  const tmp: string[] = []

  while (tmp.length < 5) tmp.push(...images)

  tmp.unshift(images[length - 1])

  activePool.value = tmp.slice(0, 5)

  if (length === 1) active.value = 0

  emit('activeChange', active.value)
}

function getSliderItemEls() {
  if (!sliderContainerEl.value) return

  sliderItemEls.push(
    ...(Array.from(
      sliderContainerEl.value.querySelectorAll(':scope > .slider-item')
    ) as HTMLDivElement[])
  )
}

function teleport() {
  sliderContainerEl.value?.removeEventListener('transitionend', teleport)

  const list = props.list,
    length = list.length,
    _active = active.value

  activePool.value = array(5, (i) => list[getIndex(_active + i - 2, length)])

  reset()

  setTimeout(() => {
    sliding = false
    if (props.autoplay) run()
  })
}

function run() {
  cancel()

  timer = setTimeout(
    () => (props.toRight ? slide('right') : slide('left')),
    DURATION
  )
}

function slide(direction: 'left' | 'right') {
  if (sliding) return

  sliding = true

  cancel()

  sliderContainerEl.value?.addEventListener('transitionend', teleport)

  const isLeft = direction === 'left'

  active.value = getIndex(active.value + (isLeft ? 1 : -1), props.list.length)

  emit('activeChange', active.value)

  sliderItemEls.forEach((el, i) => {
    transition(el, `all 300ms ease-out`)
    transform(
      el,
      +getTranslateX(el) + (isLeft ? -100 : 100),
      isLeft ? (i === 3 ? 1 : props.scale) : i === 1 ? 1 : props.scale
    )
  })
}

function slideLeft() {
  slide('left')
}

function slideRight() {
  slide('right')
}

function reset() {
  sliderItemEls.forEach((image) => transition(image, 'none'))
  sliderItemEls.forEach((image, i) =>
    transform(image, 100 * (i - 1), i === 2 ? 1 : props.scale)
  )
}

function cancel() {
  clearTimeout(timer)
}
</script>

<style scoped>
.slider-container {
  @apply children:(h-full flex-shrink-0 top-0 left-0 w-1/3 absolute center-flex);
}
</style>
