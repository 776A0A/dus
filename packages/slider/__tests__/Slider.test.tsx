import { asyncMount, fast } from '@dz7/test-utils'
import { array } from '@dz7/tools'
import { describe, expect, it, vi } from 'vitest'
import { ref } from 'vue'
import Slider from '../Slider.vue'
import { getTranslateX } from '../utils'

describe('Slider', () => {
  it('渲染正确', async () => {
    const { wrapper } = await factory({ list: generate(5) })

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.findAll('.slider-item')).toHaveLength(5)
  })

  it.each([
    [1, generate(1), array(5, () => 'a1')],
    [2, generate(2), ['a2', 'a1', 'a2', 'a1', 'a2']],
    [3, generate(3), ['a3', 'a1', 'a2', 'a3', 'a1']],
    [4, generate(4), ['a4', 'a1', 'a2', 'a3', 'a4']],
    [5, generate(5), ['a5', 'a1', 'a2', 'a3', 'a4']],
    [6, generate(6), ['a6', 'a1', 'a2', 'a3', 'a4']],
  ])('正确填充初始数据: %d', async (_i, list, expected) => {
    const { items } = await factory({ list })

    items.forEach((item, i) => expect(item.text()).toBe(expected[i]))
  })

  it('正确计算初始位置', async () => {
    const { items } = await factory({ list: generate(5) })

    expect(getTranslateX(items[0].element as HTMLElement)).toBe('-100')
    expect(getTranslateX(items[1].element as HTMLElement)).toBe('0')
    expect(getTranslateX(items[2].element as HTMLElement)).toBe('100')
    expect(getTranslateX(items[3].element as HTMLElement)).toBe('200')
    expect(getTranslateX(items[4].element as HTMLElement)).toBe('300')
  })

  it('初始化后，元素的transition为none', async () => {
    const { items } = await factory({ list: generate(5) })

    expect((items[0].element as HTMLElement).style.transition).toBe('none')
  })

  it('当autoplay为true时，自动轮播', async () => {
    const { wrapper } = await fast(() =>
      factory({ list: generate(5), autoplay: true })
    )

    expect(wrapper.vm.active).toBe(2)
  })

  it('轮播后，元素偏移正确且带有过渡动画', async () => {
    const { wrapper, getItems } = await fast(() =>
      factory({
        list: generate(5),
        autoplay: true,
      })
    )

    const items = getItems()

    expect(getTranslateX(items[0].element as HTMLElement)).toBe('-200')

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('轮播后，正确显示数据', async () => {
    const { wrapper, getItems } = await fast(() =>
      factory({
        list: generate(5),
        autoplay: true,
      })
    )

    await wrapper.trigger('transitionend')

    const items = getItems()

    expect(items[0].text()).toBe('a1')
    expect(items[1].text()).toBe('a2')
    expect(items[2].text()).toBe('a3')
    expect(items[3].text()).toBe('a4')
    expect(items[4].text()).toBe('a5')

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('向右轮播后，正确显示数据', async () => {
    const { wrapper, getItems } = await fast(() =>
      factory({
        list: generate(5),
        autoplay: true,
        toRight: true,
      })
    )

    await wrapper.trigger('transitionend')

    const items = getItems()

    expect(wrapper.vm.active).toBe(0)
    expect(items[0].text()).toBe('a4')
    expect(items[1].text()).toBe('a5')
    expect(items[2].text()).toBe('a1')
    expect(items[3].text()).toBe('a2')
    expect(items[4].text()).toBe('a3')

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('可通过按钮向左向右滑动', async () => {
    const { wrapper, leftButton, getItems, rightButton } = await factory({
      list: generate(5),
    })

    await fast(async () => {
      await leftButton.trigger('click')
      await wrapper.trigger('transitionend')
    })

    const items = getItems()

    expect(wrapper.vm.active).toBe(2)
    expect(items[0].text()).toBe('a1')
    expect(wrapper.html()).toMatchSnapshot()

    await fast(async () => {
      await rightButton.trigger('click')
      await wrapper.trigger('transitionend')
    })

    expect(wrapper.vm.active).toBe(1)
    expect(items[0].text()).toBe('a5')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('多次向右滑动后，active正确', async () => {
    const { wrapper, getItems, rightButton } = await factory({
      list: generate(5),
    })

    await fast(async () => {
      await rightButton.trigger('click')
      await wrapper.trigger('transitionend')
      vi.runAllTimers()
      await rightButton.trigger('click')
      await wrapper.trigger('transitionend')
    })

    const items = getItems()

    expect(wrapper.vm.active).toBe(4)
    expect(items[0].text()).toBe('a3')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('连续多次点击同一按钮，不会触发多次滑动', async () => {
    const { wrapper, leftButton, getItems } = await factory({
      list: generate(5),
    })

    await fast(async () => {
      await leftButton.trigger('click')
      await leftButton.trigger('click')
      await wrapper.trigger('transitionend')
    })

    const items = getItems()

    expect(wrapper.vm.active).toBe(2)
    expect(items[0].text()).toBe('a1')
  })

  it('hover到轮播时会暂停轮播', async () => {
    const { wrapper } = await fast(() =>
      factory({
        list: generate(5),
        autoplay: true,
      })
    )

    await wrapper.trigger('transitionend')
    await wrapper.trigger('mouseover')

    expect(wrapper.emitted('pause')).toBeTruthy()
  })
})

async function factory({
  list = [],
  autoplay = false,
  toRight = false,
}: {
  list?: string[]
  autoplay?: boolean
  toRight?: boolean
} = {}) {
  const outer = await asyncMount({
    setup: () => {
      const dataList = ref(list)
      const sliderIns = ref()

      return () => (
        <div>
          <button id='left' onClick={() => sliderIns.value?.slideLeft()}>
            left
          </button>
          <Slider
            list={dataList.value}
            autoplay={autoplay}
            toRight={toRight}
            ref={sliderIns}
          >
            {{
              default: ({ data }: { data: string }) => <div>{data}</div>,
            }}
          </Slider>
          <button id='right' onClick={() => sliderIns.value?.slideRight()}>
            right
          </button>
        </div>
      )
    },
  })

  const wrapper = outer.findComponent(Slider)
  const leftButton = outer.find('#left')
  const rightButton = outer.find('#right')

  const getItems = () => wrapper.findAll('.slider-item')

  const items = getItems()

  return { wrapper, getItems, items, leftButton, rightButton }
}

function generate(length: number) {
  return array(length, (i) => `a${i + 1}`)
}
