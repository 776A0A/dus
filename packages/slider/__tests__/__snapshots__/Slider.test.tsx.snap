// Vitest Snapshot v1

exports[`Slider > 向右轮播后，正确显示数据 1`] = `
"<div class=\\"flex relative overflow-hidden slider-container\\" data-v-45319cdb=\\"\\">
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(-100%, 0, 0) scale(0.8);\\">
    <div>a4</div>
  </div>
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(0%, 0, 0) scale(0.8);\\">
    <div>a5</div>
  </div>
  <div class=\\"slider-item will-change-transform slider-item__active\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(100%, 0, 0) scale(1);\\">
    <div>a1</div>
  </div>
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(200%, 0, 0) scale(0.8);\\">
    <div>a2</div>
  </div>
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(300%, 0, 0) scale(0.8);\\">
    <div>a3</div>
  </div>
</div>"
`;

exports[`Slider > 当轮播后，正确显示数据 1`] = `
"<div class=\\"flex relative overflow-hidden slider-container\\" data-v-45319cdb=\\"\\">
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(-100%, 0, 0) scale(0.8);\\">
    <div>a1</div>
  </div>
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(0%, 0, 0) scale(0.8);\\">
    <div>a2</div>
  </div>
  <div class=\\"slider-item will-change-transform slider-item__active\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(100%, 0, 0) scale(1);\\">
    <div>a3</div>
  </div>
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(200%, 0, 0) scale(0.8);\\">
    <div>a4</div>
  </div>
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(300%, 0, 0) scale(0.8);\\">
    <div>a5</div>
  </div>
</div>"
`;

exports[`Slider > 渲染正确 1`] = `
"<div class=\\"flex relative overflow-hidden slider-container\\" data-v-45319cdb=\\"\\">
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(-100%, 0, 0) scale(0.8);\\">
    <div>a5</div>
  </div>
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(0%, 0, 0) scale(0.8);\\">
    <div>a1</div>
  </div>
  <div class=\\"slider-item will-change-transform slider-item__active\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(100%, 0, 0) scale(1);\\">
    <div>a2</div>
  </div>
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(200%, 0, 0) scale(0.8);\\">
    <div>a3</div>
  </div>
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(300%, 0, 0) scale(0.8);\\">
    <div>a4</div>
  </div>
</div>"
`;

exports[`Slider > 轮播后，元素便宜正确且带有过渡动画 1`] = `
"<div class=\\"flex relative overflow-hidden slider-container\\" data-v-45319cdb=\\"\\">
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: all 300ms ease-out; transform: translate3d(-200%, 0, 0) scale(0.8);\\">
    <div>a5</div>
  </div>
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: all 300ms ease-out; transform: translate3d(-100%, 0, 0) scale(0.8);\\">
    <div>a1</div>
  </div>
  <div class=\\"slider-item will-change-transform slider-item__active\\" data-v-45319cdb=\\"\\" style=\\"transition: all 300ms ease-out; transform: translate3d(0%, 0, 0) scale(0.8);\\">
    <div>a2</div>
  </div>
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: all 300ms ease-out; transform: translate3d(100%, 0, 0) scale(1);\\">
    <div>a3</div>
  </div>
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: all 300ms ease-out; transform: translate3d(200%, 0, 0) scale(0.8);\\">
    <div>a4</div>
  </div>
</div>"
`;

exports[`Slider > 轮播后，正确显示数据 1`] = `
"<div class=\\"flex relative overflow-hidden slider-container\\" data-v-45319cdb=\\"\\">
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(-100%, 0, 0) scale(0.8);\\">
    <div>a1</div>
  </div>
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(0%, 0, 0) scale(0.8);\\">
    <div>a2</div>
  </div>
  <div class=\\"slider-item will-change-transform slider-item__active\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(100%, 0, 0) scale(1);\\">
    <div>a3</div>
  </div>
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(200%, 0, 0) scale(0.8);\\">
    <div>a4</div>
  </div>
  <div class=\\"slider-item will-change-transform\\" data-v-45319cdb=\\"\\" style=\\"transition: none; transform: translate3d(300%, 0, 0) scale(0.8);\\">
    <div>a5</div>
  </div>
</div>"
`;
